src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc

```diff
@@ -234,16 +230,24 @@ class GrpcLb : public LoadBalancingPolicy {
   class SubchannelWrapper : public DelegatingSubchannel {
    public:
     SubchannelWrapper(RefCountedPtr<SubchannelInterface> subchannel,
-                      std::string lb_token,
+                      RefCountedPtr<GrpcLb> lb_policy, std::string lb_token,
                       RefCountedPtr<GrpcLbClientStats> client_stats)
         : DelegatingSubchannel(std::move(subchannel)),
+          lb_policy_(std::move(lb_policy)),
           lb_token_(std::move(lb_token)),
           client_stats_(std::move(client_stats)) {}
 
+    ~SubchannelWrapper() override {
+      if (!lb_policy_->shutting_down_) {
+        lb_policy_->CacheDeletedSubchannel(wrapped_subchannel());
```

Comment:
We could get rid of the timer loop in the grpclb policy if we just had this dtor allocate its own object which holds a closure, and a timer which fires in `GRPC_GRPCLB_DEFAULT_SUBCHANNEL_DELETION_DELAY_MS` from now (this object would destroy itself and the unref the subchannel when the timer fired).

Reply:
We could do that, but we'd still need to keep track of all of the cached subchannels that are pending deletion, because we need to be able to cancel any pending timers when the LB policy shuts down.  So I think this would increase memory usage without any real benefit, as per my reply to Craig below.

Reply:
That would cause memory leaks on grpc shutdown.

Reply:
I think the correct behavior for any code is to cancel any async work that it has pending when it shuts down.  I would consider not doing that to be a bug.

I'm not sure what happens in the case where the subchannel has a pending TCP connection setup; it may be that we have a bug there.  But even if we do, I don't think that's a justification for adding a new bug here.